frontend/src/services/authentication_handler_service/authentication_handler_service.ts

```diff
@@ -0,0 +1,66 @@
+import { BackendAuthenticationInterface } from "../backend_service_interface/backend_service_interface";
+import { CLIENT_ID } from "./authentication_constants";
+
+/**
+ * Communicates with gapi to retrieve auth code from user.
+ * Exchanges auth code and redirect URI for ID token from backend.
+ */
+export class AuthenticationHandlerService {
+  /** Backend is responsible for retrieving the ID token. */
+  constructor(private readonly backend: BackendAuthenticationInterface) {
+    // Safe because the gapi script is loaded in the head with no async/defer.
+    // TODO: Make gapi script async and call this when it has loaded
+    window.gapi.load('auth2', () => {
+      window.gapi.auth2.init({
+        client_id: CLIENT_ID,
+      });
+    });
+  }
+
+  private async getAuthCode(
+      opt?: gapi.auth2.OfflineAccessOptions): Promise<string | undefined> {
+    const authInstance = window.gapi.auth2.getAuthInstance();
+    try {
+      const { code } = await authInstance.grantOfflineAccess(opt);
+      return code;
+    } catch (err) {
+      return;
+    }
+  }
+
+  /**
+   * Signs the user in, retrieving a token from the backend.
+   * @param scopes the scopes our app requests for oauth
+   * @param tokenConsumer a function that is called once our token is generated
+   * @param onFailure a function that is called if we fail to get a token/code
+   */
+  async signIn(
+      scopes: string,
+      tokenConsumer: (token: string) => void,
+      onFailure: () => void): Promise<void> {
```

Comment:
See if you can make this work :) The promise is already your success and failure callback

```suggestion
  async signIn(scopes: string): Promise<string> {
```

Reply:
Whoops you're absolutely correct here, the function has been changed.